kate_brownbill_dungon_fighters_v1.0.py

```diff
@@ -0,0 +1,838 @@
+        card_red = Fighter([MAGIC_RFCT, FIREBOLT, MAGIC_SWNG, SHIELD_BASH], "Red Card", "Woven from blood. Does it even exist?", 3 + i, 4 + 1, 0 + 1)
+        team.append(card_black)
+        team.append(card_red)
+    do_battle([him], team, player, foe)
+            
+menu = True
+jester = False           
+while menu == True:
+    if jester == False:
+        ans = get_menu(["dungeon crawl", "tutorial", "exit", "jevil"], "Pick a game mode\n[DUNGEON CRAWL] [TUTORIAL] [EXIT]\n", "Please pick a valid mode")
+    else:
+        ans = get_menu(["dungeon crawl", "tutorial", "exit", "jevil", "jester's game"], "Pick a game mode\n[DUNGEON CRAWL] [TUTORIAL] [JESTER'S GAME] [EXIT]\n", "Please pick a valid mode")
+    if ans == 0:
+        result = do_dungeon_crawl()            
+        if result == 1:
+            jester = True
+    elif ans == 1:
+        print("\nWelcome to dungeon fighters!")
+        print("Here you command a party of adventurers who fight various foes.")
+        print("There are 5 moves, FIREBOLT, FORWARDS SWING, MAGIC SWING")
+        print("SHIELD BASH, and MAGIC REFLECT. Every fighter has some")
+        print("combanation of moves. Once the two fighters have picked a move")
+        print("to use, the two fight. Depending on what moves they used, one")
+        print("fighter will gain the advantage in the battle and deal damage.")
+        print("\nEach fighter as 3 stats: HP ATTC DEF. HP is the hit points")
+        print("ATTC is linked directly to damage given by a fighter. DEF absorbs")
+        print("damage given by an oponent, however DEF will not compleatly absorb")
+        print("all the damage.\n")
+    elif ans == 2:
+        menu = False
+    elif ans == 3:
+        jester = True
+        print("Jester's game unlocked")
+    elif ans == 4:
+        jester_game()
+        
+
+#---------------------------UNITTESTING-------------------------
+def test_ai():
+    fighter1 = Fighter([FIREBOLT, MAGIC_SWNG], "fighter1", "a fighter", 1, 1, 1)
+    fighter2 = Fighter([SHIELD_BASH, FRWD_SWNG, MAGIC_SWNG], "fighter2", "a fighter", 1, 1, 1)
+    
+    turnA1 = agressive_ai(SHIELD_BASH, fighter1) #should return magic swing
+    turnA2 = agressive_ai(FIREBOLT, fighter2) #should return sheild bash
+    turnP1 = passive_ai(SHIELD_BASH, fighter1) #should return firebolt
+    turnP2 = passive_ai(SHIELD_BASH, fighter2) #should return sheild bash
+    turnC1 = copycat_ai(FRWD_SWNG, fighter1) #should return magic swing
+    turnC2 = copycat_ai(FRWD_SWNG, fighter2) #should return forward swing
+    
+    equal_assert(turnA1, (1, MAGIC_SWNG, fighter1))
+    equal_assert(turnA2, (1, SHIELD_BASH, fighter2))
+    equal_assert(turnP1, (1, FIREBOLT, fighter1))
+    equal_assert(turnP2, (1, SHIELD_BASH, fighter2))
+    equal_assert(turnC1, (1, MAGIC_SWNG, fighter1))
+    equal_assert(turnC2, (1, FRWD_SWNG, fighter2))
+
+def test_damage():
```

Comment:
Changing the arrangement slightly might help understanding.  i.e.,

```
 fighter1 = Fighter([FRWD_SWNG], "fighter1", "a fighter", 2, 1, 1)
 fighter1.damage(2) #state should be 1, hp should be 1
 equal_assert(fighter1.hp, 1)
 equal_assert(fighter1.state, 1)
```

Also, consider how the readability changes if I do this...

```
 fighter1 = Fighter([FRWD_SWNG], "fighter1", "a fighter", 2, 1, 1)
 fighter1.damage(2) #state should be 1, hp should be 1
 equal_assert(fighter1.hp, 1)
 equal_assert(fighter1.state, 1)
```
Everything is all together.  I can also do this..

```
 fighter1 = Fighter([FRWD_SWNG], "fighter1", "a fighter", 2, 1, 1)  # hp=2, attack=1, defense=1
 equal_assert(fighter1.hp(), 2)
 equal_assert(fighter1.state, STATE_ALIVE )
 fighter1.damage(2) #hp should be 1 (2 - defence).  Still alive
 equal_assert(fighter1.hp(), 1)
 equal_assert(fighter1.state, STATE_ALIVE )
```

It's true that a lot of unit testing guides recommend that you initialize something,  i.e.,

`fighter1 = Fighter([FRWD_SWNG], "fighter1", "a fighter", 2, 1, 1)`

Perform your action

`fighter1.damage(2) #state should be 1, hp should be 1`

and do tests...

```
    equal_assert(fighter1.hp, 1)
    equal_assert(fighter1.state, 1)
```

But this can also be seen as three separate test cases that test the damage handling code.